app/services/fact_parser.rb

```diff
@@ -33,18 +33,37 @@ def operatingsystem
   end
 
   def environment
+    Environment.unscoped.where(:name => name).first_or_create
+  end
+
+  def environment_name
     raise NotImplementedError, not_implemented_error(__method__)
   end
 
   def architecture
+    name = architecture_name
+    Architecture.where(:name => name).first_or_create if name.present?
+  end
+
+  def architecture_name
     raise NotImplementedError, not_implemented_error(__method__)
   end
 
   def model
+    name = model_name
+    Model.where(:name => name.strip).first_or_create if name.present?
+  end
+
+  def model_name
     raise NotImplementedError, not_implemented_error(__method__)
   end
 
   def domain
+    name = domain_name
+    Domain.unscoped.where(:name => name).first_or_create if name.present?
```

Comment:
the `unscoped` is scary here, by sending facts, we can create mismatches (domain from org1 while the host is in org2). I know it's not introduced in this PR, just curious if you have any thoughts how to improve this, when we're touching it

of course it applies everywhere

Reply:
Well, it was in the Puppet fact parser before. Looks like [in foreman_ansible](https://github.com/theforeman/foreman_ansible/blob/5eec05f645eae1b1cde9de2179f4cf5208873ae3/app/services/foreman_ansible/fact_parser.rb#L28-L32) it isn't. My idea was that eventually a FactParser should not interact with any DB models and only be an extract of data. By splitting it in the `domain_name` and `domain` I hope we get move in that direction. Perhaps the `domain` should be deprecated as well.

Reply:
Well the logic will need to live somewhere, perhaps we'll have some FactMapper service class at some point, the ultimate problem is, we don't know the scope for facts when facts come from the proxy.

Reply:
Agreed, but I've decided to move this out of this PR to reduce the scope. Let's move it to https://github.com/theforeman/foreman/pull/8451 so this PR can get in.